Sync manager, for issue #11

herbig opened the pull request.

Commit:
Sync manager, for issue #11

kvj commented:
Honestly, 250msec for Android application is a way too frequent. Why it's necessary to copy behaviour of web/desktop application? Sync only at onPause() call looks totally acceptable to me.

deftelf commented:
I do think we should sync periodically, it's what Evernote does and it likely will cut down on conflicts on the backend. 250ms might be a bit aggressive, but that's just a number.

I think the sync manager needs to have some code to stop it running multiple syncs. If a sync takes several seconds it could end up running many syncs simultaneously, which would be Bad.

mobitar commented:
On the iOS/desktop apps, the app syncs ever 30s. 

However, when a user is typing, it syncs after every change, or, after 250ms of keyboard idle time (if there was a change made).

Which is this issue addressing?

herbig commented:
Ah, I see.  I can do that.

So while editing a note, the app can sync with every change within 250ms.  Other than that, the app will sync every 30 seconds as well.

Should it sync remotely while editing, or just save locally, and sync when you leave the screen?  In the latter case, you wouldn't get updates if you were say looking at the web version while typing on the app, but you'd cut down on network calls considerably.

I'll also fix the multiple sync issue @deftelf mentioned.

mobitar commented:
> So while editing a note, the app can sync with every change within 250ms. Other than that, the app will sync every 30 seconds as well.

Correct

> Should it sync remotely while editing, or just save locally

Remotely. Because you don't want to count on the user leaving the screen. What if they just shut off their device screen? They'd expect to see the changes on the web app.

mobitar commented:
We should also add the sync status on top of the note. The iOS/web/desktop apps have this, and I think it really helps with peace of mind, knowing your note is saved and synced.

Here's how it looks in iOS (under "Compose"):

https://www.dropbox.com/s/tm9ixd9aais1gux/Untitled%202.mov?dl=0

herbig commented:
OK, cool. I can do that later today

On Feb 9, 2017 12:12 PM, "Mo Bitar" <notifications@github.com> wrote:

> We should also add the sync status on top of the note. The iOS/web/desktop
> apps have this, and I think it really helps with peace of mind, knowing
> your note is saved and synced.
>
> Here's how it looks in iOS (under "Compose"):
>
> https://www.dropbox.com/s/tm9ixd9aais1gux/Untitled%202.mov?dl=0
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/standardnotes/android/pull/22#issuecomment-278707721>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXeLypIzmaxIihvkzjgKtsb9SMTZjC0ks5ra0kBgaJpZM4L7pFd>
> .
>

deftelf commented:
Just to clarify, the only reason we save locally is so that we can sync to server, there is no point in saving to the disk periodically because we know we should do that when onPause() runs. Having written all that, I see a bug in the existing code where onPause() does not actually cause a save-local if you never leave the screen (so isFinishing()) will not be true.

In the pull request this isn't an issue because it's replaced by being called onDestroyView(), which has its own problems: it's not guaranteed to run, and it will run after the onResume() has run on the next Activity (eg NoteListActivity). This is probably not what we want, because the list will show the old note before it updates.

Probably we should always save after 250ms of inactivity, or if onPause() is run.

deftelf commented:
Having looked into the history I am reasonably sure that that `if (activity.isFinishing) {` shouldn't be there, it should always save

herbig commented:
Ah, yeah good call.  I'll patch that up as well, and I shouldn't be unsubscribing to changes in onDestroyView either or we'll have an activity leak..

mobitar commented:
> the only reason we save locally is so that we can sync to server

Not sure if this makes any practical difference, but some users prefer offline mode in some cases. The local copy should always match the remote.

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	app/src/main/java/org/standardnotes/notes/frag/NoteListFragment.kt

deftelf commented:
@mobitar we're talking about the same thing, just a function of Android lifecycle. We don't need to save to disk periodically in general because we can wait until the screen is hidden to do it.

herbig commented:
I could be wrong, but I think the way it's currently set up, we'd need to save to disk to sync, since the sync method grabs all dirty notes from disk.  Should I keep it that way or refactor it to do it some other way?

deftelf commented:
That's... what I mean... We've got ourselves in a circular conversation.
I'm just going to abandon this train because we seem to all agree and I'm the one not communicating effectively...

Commit:
More #11 work:
- improved sync manager
- 30s sync timer
- swapped error toast for snackbar
- sync status in toolbar

We should also start/stop the 30s timer in onResume/onPause of each activity, so it isn't running while the app is not visible. We can add that once we have a base activity

deftelf commented on app/src/main/java/org/standardnotes/notes/comms/SyncManager.kt:
```diff
@@ -1,55 +1,115 @@
     }
 
-    private val syncListeners: MutableList<SyncListener> = mutableListOf()
+    // TODO should we use a service instead?
```
I tend no unless there's a need to run outside of normal app resumed state.

deftelf commented on app/src/main/java/org/standardnotes/notes/comms/SyncManager.kt:
```diff
@@ -1,55 +1,115 @@
                 .forEach { syncListeners.remove(it) }
     }
 
     fun sync() {
```
Do we need to worry about syncrhonising this? Probably not at the moment?

herbig replied:
I don't *think* so. Everything, including network callbacks, should all be on the UI thread.

deftelf commented on app/src/main/java/org/standardnotes/notes/comms/SyncManager.kt:
```diff
@@ -1,55 +1,115 @@
+        var iter = syncListeners.iterator()
+        while (iter.hasNext()) {
+            val listening = iter.next()
+            if (listening.get() == null) {
```
Could you also add a debug log here please, we shouldn't have things in this list that have gone out of scope, there's a bug elsewhere if this is being hit

herbig replied:
Will do. My initial thought was since it's using a weakreference, we could simplify it by doing away with unsubscribe altogether.  Once the fragment/activity is destroyed it would be free for garbage collection, and the weakreference here would be null. But that feels a bit dirty to me.

deftelf replied:
Yeah that would be much pain - the reference would stay live until after onDestroyView() is called, so if any UI updates happen on the listener it would crash.

deftelf commented on app/src/main/java/org/standardnotes/notes/frag/NoteFragment.kt:
```diff
@@ -80,25 +85,25 @@ class NoteFragment : Fragment(), SyncManager.SyncListener {
     fun saveNote(note: Note?): Boolean {
-        if (note != null && (note.title != titleEdit.text.toString() ||
-                note.text != bodyEdit.text.toString())) {
+        if (note!!.title != titleEdit.text.toString() ||
```
Needs to save onPause()

deftelf commented on app/src/main/java/org/standardnotes/notes/MainActivity.kt:
```diff
@@ -25,6 +26,8 @@ class MainActivity : AppCompatActivity() {
             (supportFragmentManager.findFragmentById(R.id.noteListFrag) as NoteListFragment).startNewNote()
         }
+
+        SyncManager.startSyncTimer()
```
I think starting and stopping the sync should happen on resumed. I think this code will have the sync loops running from the MainActivity and the NoteActivity at the same time, which is unlikley to be what we want.

herbig replied:
Ah yeah good call.  I had kind of assumed it wouldn't matter that the 30s timer is also running in NoteActivity, but then you'd have a race condition where edits wouldn't get synced when typing while the global timer was syncing..

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/standardnotes/notes/MainActivity.kt
#	app/src/main/java/org/standardnotes/notes/frag/NoteListFragment.kt

Commit:
bug fixes

deftelf commented:
Looks good, cheers.
